### DISCUSSION TITLE
Cannot import name MockV3Aggregator

### DISCUSSION THREAD

**Discussion by ghost** [smartcontractkit/full-blockchain-solidity-course-py#294] [Q&A] [answered]
hello, 
around 5:31 time I was following along and pasted the MockV3Aggregator.sol from the git page and immediately it has a red error under the import line. In my FundMe.sol I have the same issue but it still works so I didn't think much. 
![mock3redling](https://user-images.githubusercontent.com/46335699/139618510-470bebbb-af9a-4c9b-844e-f583240974bc.png)

When I try to run deploy.py I get the error
ImportError: cannot import name 'MockV3Aggregator' from 'brownie' (C:\Users\Larry IV\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\__init__.py)
![mock3errorfixed](https://user-images.githubusercontent.com/46335699/139618359-c68e97e7-fe1e-43ba-87f2-4eff064a24b8.png)

in my brownie config I had redirected the atChainlink import as well. 
![browniefig](https://user-images.githubusercontent.com/46335699/139618453-46ce43d0-8dfa-4151-ab8f-72a88955f2e2.png)

**Comment by cromewar** [comment, marked as answer]
Hello @cptknvckles 

To get rid off linting issues please refer to this link: [chronological-issues-from-video.md](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/blob/main/chronological-issues-from-video.md) on Linting Issues apart, you should modify your settings.Json.

Did you compile the project again after importing the `MockV3Aggregator `? please delete the build folder an compile again with `brownie compile`.

**Reply by DarrenWayn** [reply]
<img width="877" alt="Screen Shot 2022-01-06 at 09 19 25" src="https://user-images.githubusercontent.com/22322944/148317694-3e355e53-5fe6-42c3-a1a0-14b219f28e68.png">

I still have the same problem importing MockV3Aggregator, is there anything wrong with my code?

**Reply by DarrenWayn** [reply]
<img width="758" alt="Screen Shot 2022-01-06 at 10 16 09" src="https://user-images.githubusercontent.com/22322944/148322725-d09aa8db-ba7e-4337-820e-b2cad576bd29.png">
i fix the problem but ValueError: Explorer API not set for this network, @cromewar do you know whats wrong?

**Reply by cromewar** [reply]
@DarrenWayn you must go to your brownie-config.ymal and add the info for the network you are using, adding eth_usd_price_feed: to the development network if you are testing in local.

**Reply by DarrenWayn** [reply]
Thanks it worked!!

**Comment by sl00k** [comment]
where you able to solve this? 
getting this as well now
![image](https://user-images.githubusercontent.com/33859801/143249763-2d5c8aed-3bf6-4ebb-97ab-2495363738b3.png)

**Reply by sl00k** [reply]
sorry. Just saw that I created the "test" folder with the "MockV3Aggregator" in the wrong place. Problem solved

**Reply by ktjayamanna** [reply]
same here lol

**Reply by diliprobbi** [reply]
Same here lol it must be under contracts>test>MockV3Aggregator not build>contracts>test...

**Comment by favourO** [comment]
Hi, 

I am having similar issues with mockV3Aggregator

mac@Macs-MacBook-Pro fund-me % brownie run scripts/deploy.py
Brownie v1.17.1 - Python development framework for Ethereum

  File "brownie/_cli/__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "brownie/_cli/run.py", line 41, in main
    active_project.load_config()
  File "brownie/project/main.py", line 453, in load_config
    _load_project_config(self._path)
  File "brownie/_config.py", line 222, in _load_project_config
    and "cmd_settings" in values
TypeError: argument of type 'NoneType' is not iterable

**Reply by cromewar** [reply]
It seems your brownie is having a hard time finding your ganache-cli, you should install nodeJs and Npm and install ganache-cli.

**Reply by favourO** [reply]
I already have node installed, my npm version is 8.1.0 and I just installed ganache-cli again

<img width="222" alt="Screenshot 2021-12-16 at 15 02 02" src="https://user-images.githubusercontent.com/23056250/146386274-23b8416c-0cb4-4b51-b901-6855218367b0.png">
<img width="439" alt="Screenshot 2021-12-16 at 15 02 12" src="https://user-images.githubusercontent.com/23056250/146386303-061eb2cb-1530-4ca7-ace8-c48be88cc5da.png">

**Reply by cromewar** [reply]
Looking into brownie source code: https://github.com/eth-brownie/brownie/blob/master/brownie/project/main.py
The _load_project_config seems to be missing your brownie-config.yaml, could you please make sure the name of the file is correct `brownie-config.yaml` and also check you don't have any identation error on it?

**Reply by Anu-srinivasan** [reply]
Same error for me. When I am running scripts for NFT lesson.

Brownie v1.16.4 - Python development framework for Ethereum

  File "C:\Users\RBI\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "C:\Users\RBI\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\run.py", line 40, in main     
    active_project.load_config()
  File "C:\Users\RBI\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 452, in load_config
    _load_project_config(self._path)
  File "C:\Users\RBI\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_config.py", line 222, in _load_project_config
    and "cmd_settings" in values
TypeError: argument of type 'NoneType' is not iterable



Checked indentation in brownie-config.yaml and spelling of brownie-config.yaml file. They are perfect.  Please help

**Reply by cromewar** [reply]
could you try using pip for brownie instead of pipx?

**Reply by Ugwumsi** [reply]
I have the same issue as FavourO above and my config file is named properly and without any indentation errors, but I still get the same error message. What else could be the cause?
![Screenshot 2022-02-05 at 18 17 35](https://user-images.githubusercontent.com/81196037/152652186-980b9fe9-f17f-4fa0-90f3-558229636f09.png)

**Reply by cromewar** [reply]
@Egbu Please share more info on this:

1. the command you are using to deploy de contract.
2. have you added the network correctly into brownie?
3. your helpful_scripts as well.

**Reply by Ugwumsi** [reply]
1. I use 'brownie run scripts/deploy.py' to deploy the contract.
2. Here is a screenshot of the function that selects an account depending on the network I'm on.
![Screenshot 2022-02-09 at 15 00 53](https://user-images.githubusercontent.com/81196037/153216637-21c06d16-9c89-4ac4-86f0-5a5e7bd62c0c.png)

**Comment by codeforstartups** [comment]
Facing the same issue. 

![image](https://user-images.githubusercontent.com/48567820/148954089-599eb7cc-de7c-4997-8b81-0c62b176eb5c.png)

**Reply by cromewar** [reply]
Hello @codeforstartups please share your project folder structure here.

**Comment by codeforstartups** [comment]
@cromewar 

![image](https://user-images.githubusercontent.com/48567820/148961184-5a3d06f5-eb0e-4ff5-ab19-d0c337447767.png)

**Reply by cromewar** [reply]
Your MockV3 aggregator should be on a folder called "test" inside your contracts folder, not in the "tests" folder which is meant to your python test scripts.

**Comment by codeforstartups** [comment]
ok, let me check

**Comment by codeforstartups** [comment]
@cromewar did the same, 
![image](https://user-images.githubusercontent.com/48567820/148967226-3d32d76c-5ca5-44bb-95fc-dd7593dbb383.png)

**Reply by cromewar** [reply]
after moving the folder, you should delete the `build` folder and compile again.

**Reply by cromewar** [reply]
Also is saying you are missing a `;` on the mock weth contract.

**Comment by codeforstartups** [comment]
@cromewar but it did not worked

**Comment by codeforstartups** [comment]
@cromewar , I did "after moving the folder, you should delete the build folder and compile again." , but still no luck
![image](https://user-images.githubusercontent.com/48567820/149059589-6e94f3a7-b147-4362-8c74-03e43c2ba2d2.png)

**Comment by codeforstartups** [comment]
I m calling MockV3Aggregator inside of helpful_scripts , got from chainlink, 

![image](https://user-images.githubusercontent.com/48567820/149059960-504962df-5bcc-448a-8062-5deb195a5d46.png)

**Comment by codeforstartups** [comment]
Solution 👍 

Steps : 
1. MockV3Aggregator File inside contracts/test/MockV3Aggregator.sol 
2. brownie compile --all
3. brownie run scripts/deploy.py  
4. Deployed Successful

**Reply by cromewar** [reply]
I'm glad is solved now

**Reply by bubu59** [reply]
Hi guys, I have done all the steps as mentioned in the discussion but am still facing the same error. Please see the attached picture below. Appreciate any help! 
![image](https://user-images.githubusercontent.com/74093051/159145683-932dab98-7407-4815-9835-5d6fbf85bd82.png)

**Reply by cromewar** [reply]
Change the folder name to tests with S and try compiling again.

**Reply by bubu59** [reply]
Still getting the same error. Did I change it right? 
![image](https://user-images.githubusercontent.com/74093051/159145889-72df7eab-1ee6-41ce-b06b-63397393ac83.png)

**Reply by cromewar** [reply]
Yest is ok, which is the output of `brownie compile`?

**Reply by bubu59** [reply]
Here you go
![image](https://user-images.githubusercontent.com/74093051/159146014-4de20fb4-c0d7-4c3e-b819-aa104a252829.png)

**Reply by cromewar** [reply]
Interesting I know what the problem is now, as you can see your MockV3Aggragator have a contract name `MockAggregator` without the v3, and you are trying to import using the V3 identifier, please delete the Mock and copy it again correctly, remember the name of the contract (no the file) should be `MockV3Aggregator`.

**Reply by bubu59** [reply]
Oh, for that I'm using v0.8 that's why the contract is MockAggregator. 
![image](https://user-images.githubusercontent.com/74093051/159146510-3cd44b38-72d3-456a-9e4b-10a6234c9b00.png)

**Reply by bubu59** [reply]
I also went to correct it in the deploy.py as well 
![image](https://user-images.githubusercontent.com/74093051/159146533-0e66f805-b817-4976-8959-8e75e78d621a.png)

**Reply by bubu59** [reply]
However, when I went to run deploy.py again, it says now I have a key error. I have checked against the spelling in my brownie-config.yaml folder and its the same spelling. 
![image](https://user-images.githubusercontent.com/74093051/159146639-d2d8c859-9157-41e8-9bdf-ed0528a6100f.png)
![image](https://user-images.githubusercontent.com/74093051/159146643-983b9938-a1d1-41a4-8152-62c4b10e7e72.png)

**Reply by cromewar** [reply]
Well It depends on which network you are deploying, as you can see you just have the eth_usd_price_feed on kovan, so you must add it to any other network you want to deploy on, like rinkeby or development.

**Reply by bubu59** [reply]
okkk... But in this instance when I ran the command brownie run scripts/deploy.py, I did not state any network flag, which should point it towards the development chain? so why should I add the price feed which is for the kovan testnet to the other networks, ie development?

**Reply by cromewar** [reply]
the default flag is development, you can use the same one as on kovan or rinkeby as you are going to deploy mock anyways but brownie is going to complain if it does not find it.

**Reply by bubu59** [reply]
ok. So I have made the changes but now its saying that Web3 is not connected.. 
![image](https://user-images.githubusercontent.com/74093051/159147818-a6f2f18a-5abd-46e5-829a-bf44eff2d5b2.png)

**Reply by cromewar** [reply]
make sure you are using ganache correctly.

**Reply by bubu59** [reply]
ooh. What do you mean use ganache correctly? I have managed to get it running but when I re-run deploy.py it crashes and gives the same error.
![image](https://user-images.githubusercontent.com/74093051/159153008-8e75eec6-83a8-4140-951e-077a773ac949.png)

**Reply by cromewar** [reply]
Be sure to have ganache hi running and be sure your ganache-cli is also installed correctly

**Reply by bubu59** [reply]
ahhh... yes, it runs now! Thank you! However, the MockAggregator is not deployed. Any thoughts why that might be the case? Thank you!
![image](https://user-images.githubusercontent.com/74093051/159197786-a98f3d8c-9492-4af2-871e-dc12aaae48ca.png)

**Comment by nameless-codingman** [comment]
im the same problem ,i dont know why  ,the same name could run successful  
![image](https://user-images.githubusercontent.com/67813780/219079982-abbcae5e-f05b-4141-8387-190ee34150f1.png)
![image](https://user-images.githubusercontent.com/67813780/219080026-e8c5c583-6bf0-455a-9bd4-e7a0da67333d.png)
![image](https://user-images.githubusercontent.com/67813780/219080225-ac9d2789-7a42-44c3-9424-f1f60871a6b0.png)

**Reply by cromewar** [reply]
Try deleting the build folder and try again.
